pkg/agent/plugin/nodeattestor/devid/tpm/signingkey.go

```diff
@@ -0,0 +1,168 @@
+		eccParams := pub.ECCParameters
+		if eccParams != nil {
+			sigHashAlg = eccParams.Sign.Hash
+		}
+
+	default:
+		return nil, fmt.Errorf("bad key type: 0x%04x", pub.Type)
+	}
+
+	if sigHashAlg.IsNull() {
+		return nil, fmt.Errorf("signature hash algorithm is NULL")
+	}
+
+	srk, err := tpm2tools.NewKey(rw, tpm2.HandleOwner, srkTemplate)
+	if err != nil {
+		return nil, fmt.Errorf("SRK creation failed: %w", err)
+	}
+
+	defer srk.Close()
+
+	keyHandle, _, err := tpm2.Load(rw, srk.Handle(), "", pubKey, privKey)
+	if err != nil {
+		return nil, fmt.Errorf("tpm2.Load failed: %w", err)
+	}
+
+	return &SigningKey{
+		Handle:     keyHandle,
+		sigHashAlg: sigHashAlg,
+		rw:         rw,
+	}, nil
+}
+
+// Close removes the key from the TPM
+func (ac *SigningKey) Close() error {
+	return tpm2.FlushContext(ac.rw, ac.Handle)
+}
+
+// Sign request the TPM to sign the given data using this key
+func (ac *SigningKey) Sign(data []byte) ([]byte, error) {
+	digest, token, err := tpm2.Hash(ac.rw, ac.sigHashAlg, data, tpm2.HandlePlatform)
+	if err != nil {
+		return nil, fmt.Errorf("tpm2.Hash failed: %w", err)
+	}
+
+	var sig *tpm2.Signature
+	for {
+		sig, err = tpm2.Sign(ac.rw, ac.Handle, "", digest, token, nil)
+		if err != nil {
+			if isRetry(err) {
+				continue
+			}
+
+			return nil, fmt.Errorf("tpm2.Sign failed: %w", err)
+		}
+
+		break
+	}
+
+	if sig.RSA != nil {
+		return sig.RSA.Signature, nil
+	}
+
+	if sig.ECC != nil {
+		var b cryptobyte.Builder
+		b.AddASN1(asn1.SEQUENCE, func(b *cryptobyte.Builder) {
+			b.AddASN1BigInt(sig.ECC.R)
+			b.AddASN1BigInt(sig.ECC.S)
+		})
+
+		return b.Bytes()
+	}
+
+	return nil, fmt.Errorf("bad tpm2.Signature")
+}
+
+// Certify calls tpm2.Certify using the current key as signer and the provided
+// object as object handle.
+func (ac *SigningKey) Certify(object tpmutil.Handle) ([]byte, []byte, error) {
+	// For some reason 'tpm2.Certify()' sometimes fails the first attempt and asks for retry.
+	// So, we retry some times in case of getting the RCRetry error.
+	// It seems that this issue has been reported: https://github.com/google/go-tpm/issues/59
+	var certifiedDevID []byte
+	var certificationSignature []byte
+	var err error
+	for i := 0; i <= maxAttempts; i++ {
+		certifiedDevID, certificationSignature, err = tpm2.Certify(ac.rw, "", "", object, ac.Handle, nil)
+		switch {
+		case err == nil:
+			return certifiedDevID, certificationSignature, nil
+
+		case isRetry(err):
+			if i == maxAttempts {
+				return nil, nil, fmt.Errorf("max attempts reached: %w", err)
+			}
+
+			log.Printf("TPM was not able to start the command. Retrying: attempt (%d/%d)", i, maxAttempts)
+			time.Sleep(time.Millisecond * 500)
+
+		default:
+			break
+		}
+	}
+
+	return nil, nil, fmt.Errorf("certify failed: %w", err)
+}
+
+// SRKTemplateRSA returns SRK template used during provisioning.
+func SRKTemplateRSA() tpm2.Public {
+	template := tpm2tools.SRKTemplateRSA()
+	template.RSAParameters.ModulusRaw = []byte{}
```

Comment:
The gotpm library does not have a method to create the default SRK template recommended by the latest version of the specification. So we create the template on our side. I’ll add a comment and rename the function to make this clearer.


According to [TCG TPM v2.0 Provisioning Guidance (section 7.5.1)](https://trustedcomputinggroup.org/resource/tcg-tpm-v2-0-provisioning-guidance/) the Storage Root Key (SRK) template should be based on the Endorsement Key (EK) template.
The EK template is defined in [TCG EK Credential Profile for TPM Family 2.0 (section B)](https://trustedcomputinggroup.org/resource/tcg-ek-credential-profile-for-tpm-family-2-0/). The latest version of this document (published in july 2020) defined a new template (called H-1) for the RSA EK template.
The new template sets the unique field (named here as ModulusRaw) to an empty buffer instead of a well-known buffer (filled with 0s).